// Plantation Model Service - gRPC API Definition
// Master data registry for the Farmer Power Platform

syntax = "proto3";

package farmer_power.plantation.v1;

import "google/protobuf/timestamp.proto";

// Note: Health checking uses standard grpc.health.v1.Health (grpc-health-checking package)
// No custom HealthService defined here to avoid duplication.

// ============================================================================
// Plantation Service - Core Operations
// ============================================================================

service PlantationService {
  // Region operations
  rpc GetRegion(GetRegionRequest) returns (Region);
  rpc ListRegions(ListRegionsRequest) returns (ListRegionsResponse);
  rpc CreateRegion(CreateRegionRequest) returns (Region);
  rpc UpdateRegion(UpdateRegionRequest) returns (Region);

  // Factory operations
  rpc GetFactory(GetFactoryRequest) returns (Factory);
  rpc ListFactories(ListFactoriesRequest) returns (ListFactoriesResponse);
  rpc CreateFactory(CreateFactoryRequest) returns (Factory);
  rpc UpdateFactory(UpdateFactoryRequest) returns (Factory);
  rpc DeleteFactory(DeleteFactoryRequest) returns (DeleteFactoryResponse);

  // Farmer operations
  rpc GetFarmer(GetFarmerRequest) returns (Farmer);
  rpc GetFarmerByPhone(GetFarmerByPhoneRequest) returns (Farmer);
  rpc ListFarmers(ListFarmersRequest) returns (ListFarmersResponse);
  rpc CreateFarmer(CreateFarmerRequest) returns (Farmer);
  rpc UpdateFarmer(UpdateFarmerRequest) returns (Farmer);

  // Collection Point operations
  rpc GetCollectionPoint(GetCollectionPointRequest) returns (CollectionPoint);
  rpc ListCollectionPoints(ListCollectionPointsRequest) returns (ListCollectionPointsResponse);
  rpc CreateCollectionPoint(CreateCollectionPointRequest) returns (CollectionPoint);
  rpc UpdateCollectionPoint(UpdateCollectionPointRequest) returns (CollectionPoint);
  rpc DeleteCollectionPoint(DeleteCollectionPointRequest) returns (DeleteCollectionPointResponse);

  // Performance summary operations
  rpc GetPerformanceSummary(GetPerformanceSummaryRequest) returns (PerformanceSummary);

  // Grading Model operations
  rpc CreateGradingModel(CreateGradingModelRequest) returns (GradingModel);
  rpc GetGradingModel(GetGradingModelRequest) returns (GradingModel);
  rpc GetFactoryGradingModel(GetFactoryGradingModelRequest) returns (GradingModel);
  rpc AssignGradingModelToFactory(AssignGradingModelToFactoryRequest) returns (GradingModel);

  // Farmer Performance operations
  rpc GetFarmerSummary(GetFarmerSummaryRequest) returns (FarmerSummary);

  // Communication Preferences operations (Story 1.5)
  rpc UpdateCommunicationPreferences(UpdateCommunicationPreferencesRequest) returns (UpdateCommunicationPreferencesResponse);
}

// ============================================================================
// Common Types
// ============================================================================

message GeoLocation {
  double latitude = 1;
  double longitude = 2;
  double altitude_meters = 3;
}

message ContactInfo {
  string phone = 1;
  string email = 2;
  string address = 3;
}

// Factory-configurable quality thresholds for farmer categorization (Story 1.7)
// NEUTRAL NAMING: tier_1, tier_2, tier_3 (NOT WIN/WATCH/WORK)
// Engagement Model maps these to engagement categories.
message QualityThresholds {
  double tier_1 = 1;  // Premium tier threshold (default ≥85% Primary)
  double tier_2 = 2;  // Standard tier threshold (default ≥70% Primary)
  double tier_3 = 3;  // Acceptable tier threshold (default ≥50% Primary)
}

// ============================================================================
// Region Messages
// ============================================================================

message Region {
  string id = 1;
  string name = 2;
  string code = 3;
  GeoLocation center = 4;
  string parent_region_id = 5;
  bool is_active = 6;
  google.protobuf.Timestamp created_at = 7;
  google.protobuf.Timestamp updated_at = 8;
}

message GetRegionRequest {
  string id = 1;
}

message ListRegionsRequest {
  int32 page_size = 1;
  string page_token = 2;
  string parent_region_id = 3;
  bool active_only = 4;
}

message ListRegionsResponse {
  repeated Region regions = 1;
  string next_page_token = 2;
  int32 total_count = 3;
}

message CreateRegionRequest {
  string name = 1;
  string code = 2;
  GeoLocation center = 3;
  string parent_region_id = 4;
}

message UpdateRegionRequest {
  string id = 1;
  optional string name = 2;
  optional string code = 3;
  optional GeoLocation center = 4;
  optional bool is_active = 5;
}

// ============================================================================
// Factory Messages
// ============================================================================

message Factory {
  string id = 1;
  string name = 2;
  string code = 3;
  string region_id = 4;
  GeoLocation location = 5;
  ContactInfo contact = 6;
  int32 processing_capacity_kg = 7;
  QualityThresholds quality_thresholds = 8;  // Story 1.7: Quality tier thresholds
  bool is_active = 9;
  google.protobuf.Timestamp created_at = 10;
  google.protobuf.Timestamp updated_at = 11;
}

message GetFactoryRequest {
  string id = 1;
}

message ListFactoriesRequest {
  int32 page_size = 1;
  string page_token = 2;
  string region_id = 3;
  bool active_only = 4;
}

message ListFactoriesResponse {
  repeated Factory factories = 1;
  string next_page_token = 2;
  int32 total_count = 3;
}

message CreateFactoryRequest {
  string name = 1;
  string code = 2;
  string region_id = 3;
  GeoLocation location = 4;
  ContactInfo contact = 5;
  int32 processing_capacity_kg = 6;
  QualityThresholds quality_thresholds = 7;  // Optional, defaults to 85/70/50
}

message UpdateFactoryRequest {
  string id = 1;
  optional string name = 2;
  optional string code = 3;
  optional GeoLocation location = 4;
  optional ContactInfo contact = 5;
  optional int32 processing_capacity_kg = 6;
  optional QualityThresholds quality_thresholds = 7;  // Story 1.7
  optional bool is_active = 8;
}

message DeleteFactoryRequest {
  string id = 1;
}

message DeleteFactoryResponse {
  bool success = 1;
}

// ============================================================================
// Collection Point Messages
// ============================================================================

message OperatingHours {
  string weekdays = 1;   // e.g., "06:00-10:00"
  string weekends = 2;   // e.g., "07:00-09:00"
}

message CollectionPointCapacity {
  int32 max_daily_kg = 1;
  string storage_type = 2;       // "covered_shed", "open_air", "refrigerated"
  bool has_weighing_scale = 3;
  bool has_qc_device = 4;
}

message CollectionPoint {
  string id = 1;                                    // Format: {region}-cp-XXX
  string name = 2;
  string factory_id = 3;
  GeoLocation location = 4;
  string region_id = 5;
  string clerk_id = 6;
  string clerk_phone = 7;
  OperatingHours operating_hours = 8;
  repeated string collection_days = 9;             // ["mon", "wed", "fri", "sat"]
  CollectionPointCapacity capacity = 10;
  string status = 11;                              // "active", "inactive", "seasonal"
  google.protobuf.Timestamp created_at = 12;
  google.protobuf.Timestamp updated_at = 13;
}

message GetCollectionPointRequest {
  string id = 1;
}

message ListCollectionPointsRequest {
  int32 page_size = 1;
  string page_token = 2;
  string factory_id = 3;
  string region_id = 4;
  string status = 5;
  bool active_only = 6;
}

message ListCollectionPointsResponse {
  repeated CollectionPoint collection_points = 1;
  string next_page_token = 2;
  int32 total_count = 3;
}

message CreateCollectionPointRequest {
  string name = 1;
  string factory_id = 2;
  GeoLocation location = 3;
  string region_id = 4;
  string clerk_id = 5;
  string clerk_phone = 6;
  OperatingHours operating_hours = 7;
  repeated string collection_days = 8;
  CollectionPointCapacity capacity = 9;
  string status = 10;
}

message UpdateCollectionPointRequest {
  string id = 1;
  optional string name = 2;
  optional string clerk_id = 3;
  optional string clerk_phone = 4;
  optional OperatingHours operating_hours = 5;
  repeated string collection_days = 6;
  optional CollectionPointCapacity capacity = 7;
  optional string status = 8;
}

message DeleteCollectionPointRequest {
  string id = 1;
}

message DeleteCollectionPointResponse {
  bool success = 1;
}

// ============================================================================
// Farmer Messages
// ============================================================================

// Farm scale classification based on hectares
enum FarmScale {
  FARM_SCALE_UNSPECIFIED = 0;
  FARM_SCALE_SMALLHOLDER = 1;   // < 1 hectare
  FARM_SCALE_MEDIUM = 2;        // 1-5 hectares
  FARM_SCALE_ESTATE = 3;        // > 5 hectares
}

// Channel for pushing notifications to farmers (Story 1.5)
// Distinct from InteractionPreference which controls consumption mode
enum NotificationChannel {
  NOTIFICATION_CHANNEL_UNSPECIFIED = 0;
  NOTIFICATION_CHANNEL_SMS = 1;       // Text messages via Africa's Talking (default)
  NOTIFICATION_CHANNEL_WHATSAPP = 2;  // WhatsApp messages
}

// Preferred mode for consuming detailed information (Story 1.5)
// Even voice-preferring farmers receive SMS triggers; they call IVR to listen
enum InteractionPreference {
  INTERACTION_PREFERENCE_UNSPECIFIED = 0;
  INTERACTION_PREFERENCE_TEXT = 1;   // Prefers reading SMS/WhatsApp (default)
  INTERACTION_PREFERENCE_VOICE = 2;  // Prefers listening via IVR/Voice AI
}

// Preferred language for farmer communications (Story 1.5)
enum PreferredLanguage {
  PREFERRED_LANGUAGE_UNSPECIFIED = 0;
  PREFERRED_LANGUAGE_SW = 1;   // Swahili (default)
  PREFERRED_LANGUAGE_KI = 2;   // Kikuyu
  PREFERRED_LANGUAGE_LUO = 3;  // Luo
  PREFERRED_LANGUAGE_EN = 4;   // English
}

message Farmer {
  string id = 1;                                    // Format: WM-XXXX (e.g., WM-0001)
  string grower_number = 2;                         // External/legacy ID if any
  string first_name = 3;
  string last_name = 4;
  string region_id = 5;                             // Auto-assigned based on GPS + altitude
  string collection_point_id = 6;                   // Primary registration collection point
  GeoLocation farm_location = 7;
  ContactInfo contact = 8;
  double farm_size_hectares = 9;
  FarmScale farm_scale = 10;                        // Auto-calculated from farm_size_hectares
  string national_id = 11;                          // Government ID
  google.protobuf.Timestamp registration_date = 12;
  bool is_active = 13;
  google.protobuf.Timestamp created_at = 14;
  google.protobuf.Timestamp updated_at = 15;
  NotificationChannel notification_channel = 16;    // Notification delivery channel (Story 1.5)
  InteractionPreference interaction_pref = 17;      // Information consumption mode (Story 1.5)
  PreferredLanguage pref_lang = 18;                 // Language preference (Story 1.5)
}

message GetFarmerRequest {
  string id = 1;
}

message GetFarmerByPhoneRequest {
  string phone = 1;
}

message ListFarmersRequest {
  int32 page_size = 1;
  string page_token = 2;
  string region_id = 3;
  string collection_point_id = 4;
  bool active_only = 5;
}

message ListFarmersResponse {
  repeated Farmer farmers = 1;
  string next_page_token = 2;
  int32 total_count = 3;
}

message CreateFarmerRequest {
  string first_name = 1;
  string last_name = 2;
  string collection_point_id = 3;                   // Primary registration collection point
  GeoLocation farm_location = 4;
  ContactInfo contact = 5;
  double farm_size_hectares = 6;
  string national_id = 7;                           // Government ID (required)
  string grower_number = 8;                         // External/legacy ID (optional)
}

message UpdateFarmerRequest {
  string id = 1;
  optional string first_name = 2;
  optional string last_name = 3;
  optional GeoLocation farm_location = 4;
  optional ContactInfo contact = 5;
  optional double farm_size_hectares = 6;
  optional bool is_active = 7;
}

// ============================================================================
// Performance Summary Messages
// ============================================================================

message PerformanceSummary {
  string id = 1;
  string entity_type = 2;  // "farmer", "factory", "region"
  string entity_id = 3;
  string period = 4;  // "daily", "weekly", "monthly", "yearly"
  google.protobuf.Timestamp period_start = 5;
  google.protobuf.Timestamp period_end = 6;

  // Aggregated metrics
  double total_green_leaf_kg = 7;
  double total_made_tea_kg = 8;
  int32 collection_count = 9;
  double average_quality_score = 10;

  google.protobuf.Timestamp created_at = 11;
  google.protobuf.Timestamp updated_at = 12;
}

message GetPerformanceSummaryRequest {
  string entity_type = 1;
  string entity_id = 2;
  string period = 3;
  google.protobuf.Timestamp period_start = 4;
}

// ============================================================================
// Grading Model Messages
// ============================================================================

// Grading type classification
enum GradingType {
  GRADING_TYPE_UNSPECIFIED = 0;
  GRADING_TYPE_BINARY = 1;       // Accept/Reject (Primary/Secondary)
  GRADING_TYPE_TERNARY = 2;      // Premium/Standard/Reject
  GRADING_TYPE_MULTI_LEVEL = 3;  // A/B/C/D or custom levels
}

// Trend direction for performance metrics
enum TrendDirection {
  TREND_DIRECTION_UNSPECIFIED = 0;
  TREND_DIRECTION_IMPROVING = 1;
  TREND_DIRECTION_STABLE = 2;
  TREND_DIRECTION_DECLINING = 3;
}

// Definition of a single attribute in the grading model
message GradingAttribute {
  int32 num_classes = 1;           // Number of classes for this attribute
  repeated string classes = 2;      // Class labels in order
}

// Conditional rejection rule
message ConditionalReject {
  string if_attribute = 1;          // Attribute to check first
  string if_value = 2;              // Value that triggers the condition
  string then_attribute = 3;        // Second attribute to check
  repeated string reject_values = 4; // Values that cause rejection
}

// Rules for determining final grade from attributes
message GradeRules {
  map<string, StringList> reject_conditions = 1;  // Always-reject attribute values
  repeated ConditionalReject conditional_reject = 2;
}

// Helper message for map of string lists (protobuf doesn't support map<string, repeated string>)
message StringList {
  repeated string values = 1;
}

// Complete grading model definition
message GradingModel {
  string model_id = 1;
  string model_version = 2;
  string regulatory_authority = 3;  // Optional: regulatory body
  string crops_name = 4;
  string market_name = 5;
  GradingType grading_type = 6;
  map<string, GradingAttribute> attributes = 7;
  GradeRules grade_rules = 8;
  map<string, string> grade_labels = 9;  // Internal grade → display label
  repeated string active_at_factory = 10;
  google.protobuf.Timestamp created_at = 11;
  google.protobuf.Timestamp updated_at = 12;
}

message GetGradingModelRequest {
  string model_id = 1;
}

message GetFactoryGradingModelRequest {
  string factory_id = 1;
}

message CreateGradingModelRequest {
  string model_id = 1;
  string model_version = 2;
  string regulatory_authority = 3;
  string crops_name = 4;
  string market_name = 5;
  GradingType grading_type = 6;
  map<string, GradingAttribute> attributes = 7;
  GradeRules grade_rules = 8;
  map<string, string> grade_labels = 9;
  repeated string active_at_factory = 10;
}

message AssignGradingModelToFactoryRequest {
  string model_id = 1;
  string factory_id = 2;
}

// ============================================================================
// Farmer Performance Messages
// ============================================================================

// Distribution counts (for map support)
message DistributionCounts {
  map<string, int32> counts = 1;
}

// Historical performance metrics (computed by batch jobs)
message HistoricalMetrics {
  // Grade-level distributions
  map<string, int32> grade_distribution_30d = 1;
  map<string, int32> grade_distribution_90d = 2;
  map<string, int32> grade_distribution_year = 3;

  // Attribute-level distributions (attribute → class → count)
  map<string, DistributionCounts> attribute_distributions_30d = 4;
  map<string, DistributionCounts> attribute_distributions_90d = 5;
  map<string, DistributionCounts> attribute_distributions_year = 6;

  // Derived convenience metrics
  double primary_percentage_30d = 7;
  double primary_percentage_90d = 8;
  double primary_percentage_year = 9;

  // Volume metrics
  double total_kg_30d = 10;
  double total_kg_90d = 11;
  double total_kg_year = 12;

  // Yield metrics
  double yield_kg_per_hectare_30d = 13;
  double yield_kg_per_hectare_90d = 14;
  double yield_kg_per_hectare_year = 15;

  // Trend
  TrendDirection improvement_trend = 16;
  google.protobuf.Timestamp computed_at = 17;
}

// Today's performance metrics (updated by streaming events)
message TodayMetrics {
  int32 deliveries = 1;
  double total_kg = 2;
  map<string, int32> grade_counts = 3;
  map<string, DistributionCounts> attribute_counts = 4;
  google.protobuf.Timestamp last_delivery = 5;
  string metrics_date = 6;  // ISO date string (YYYY-MM-DD)
}

// Farmer summary with performance details (AC #4)
message FarmerSummary {
  string farmer_id = 1;

  // Farmer basic info
  string first_name = 2;
  string last_name = 3;
  string phone = 4;
  string collection_point_id = 5;
  double farm_size_hectares = 6;
  FarmScale farm_scale = 7;

  // Grading model reference
  string grading_model_id = 8;
  string grading_model_version = 9;

  // Performance metrics
  HistoricalMetrics historical = 10;
  TodayMetrics today = 11;

  // Overall trend (convenience field)
  TrendDirection trend_direction = 12;

  google.protobuf.Timestamp created_at = 13;
  google.protobuf.Timestamp updated_at = 14;

  // Communication preferences (Story 1.5)
  NotificationChannel notification_channel = 15;
  InteractionPreference interaction_pref = 16;
  PreferredLanguage pref_lang = 17;
}

message GetFarmerSummaryRequest {
  string farmer_id = 1;
}

// ============================================================================
// Communication Preferences Messages (Story 1.5)
// ============================================================================

message UpdateCommunicationPreferencesRequest {
  string farmer_id = 1;
  NotificationChannel notification_channel = 2;
  InteractionPreference interaction_pref = 3;
  PreferredLanguage pref_lang = 4;
}

message UpdateCommunicationPreferencesResponse {
  Farmer farmer = 1;
}
